code/modules/logging/log_holder.dm

```diff
@@ -22,11 +21,102 @@ GLOBAL_REAL(logger, /datum/log_holder)
 	/// Whether or not logging as human readable text is enabled
 	var/human_readable_enabled = FALSE
 
+	/// Cached ui_data
+	var/list/data_cache = list()
+
+	/// Last time the ui_data was updated
+	var/last_data_update = 0
+
 	var/initialized = FALSE
 	var/shutdown = FALSE
 
 GENERAL_PROTECT_DATUM(/datum/log_holder)
 
+/client/proc/log_viewer_new()
+	set name = "View Round Logs"
+	set category = "Admin"
+	logger.ui_interact(mob)
+
+/datum/log_holder/ui_interact(mob/user, datum/tgui/ui)
+	if(!check_rights_for(user.client, R_ADMIN))
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(isnull(ui))
+		ui = new(user, src, "LogViewer")
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/datum/log_holder/ui_state(mob/user)
+	return GLOB.admin_state
+
+/datum/log_holder/ui_static_data(mob/user)
+	var/list/data = list(
+		"round_id" = GLOB.round_id,
+		"logging_start_timestamp" = logging_start_timestamp,
+		)
+
+	var/list/tree = list()
+	data["tree"] = tree
+	var/list/enabled_categories = list()
+	for(var/enabled in log_categories)
+		enabled_categories += enabled
+	tree["enabled"] = enabled_categories
+
+	var/list/disabled_categories = list()
+	for(var/disabled in src.disabled_categories)
+		disabled_categories += disabled
+	tree["disabled"] = disabled_categories
+
+	return data
+
+/datum/log_holder/ui_data(mob/user)
+	if(!last_data_update || (world.time - last_data_update) > LOG_UPDATE_TIMEOUT)
+		cache_ui_data()
+	return data_cache
+
+/datum/log_holder/proc/cache_ui_data()
+	var/list/category_map = list()
+	for(var/datum/log_category/category as anything in log_categories)
+		category = log_categories[category]
+		var/list/category_data = list()
+		category_data[category.category] = category_data
+
+		var/list/entries = list()
+		for(var/datum/log_entry/entry as anything in category.entries)
+			entries += list(list(
+				"id" = entry.id,
+				"message" = entry.message,
+				"timestamp" = entry.timestamp,
+				"data" = entry.data,
+				"semver" = entry.semver_store,
+				))
+		category_data["entries"] = entries
+		category_data["entry_count"] = category.entry_count
+
+		category_map[category.category] = category_data
+
+	data_cache.Cut()
+	last_data_update = world.time
+
+	data_cache["categories"] = category_map
+	data_cache["last_data_update"] = last_data_update
+
+/datum/log_holder/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("re-render")
+			log_admin("[key_name(usr)] forced a log re-render.", list("last_data_update" = last_data_update))
```

Comment:
do we really need to log this? this seems un-necessary

Reply:
```suggestion
```

Reply:
was meant to be a normal update